Feature/sc 26865/sheets desktop topics display sheets only

stevekaplan123 opened the pull request:
## Description
This PR implements a requirement for modularization -- while in the 'sheets' module, topic pages should only display sheet sources rather than library sources.

## Code Changes
1. I re-factored the TopicPage component to make the component smaller and easier to read.  I pulled out a lot of functionality into a new component TopicPageTabView which wrap the existing component TabView.
2. processTopicsTabsData -- This is the substantive change for this PR.  This function gets called whenever the topic API is called.  This function makes sure that the Sefaria._topics cache entry for the given topic contains the appropriate tabs.  In the case of the 'sheets' module, this function will now put the topics' ref links in the 'sheets' tab in the cache, whereas in the 'library' module, the refs will be stored under the 'sources' or 'notable-sources' tab.  For example, suppose I go to the library today to the topic "Pharoah".  Sefaria._topics Pharoah entry will look like this with 186 refs in the 'sources' tab:
<img width="980" alt="image" src="https://github.com/user-attachments/assets/f5587f46-1ee9-4e7a-bc3c-5901491d905b" />
In order to do this, I created a helper function _createTabKeyAndTitle to simplify the code.

YishaiGlasner commented on static/css/s2.css:
```diff
@@ -2784,7 +2784,9 @@ a.cardTitle:hover {
-.interface-hebrew .navBlockDescription.clamped .int-he {
+.interface-hebrew .navBlockDescription.clamped .int-he,
+.cardDescription
+{
```
remove newline

YishaiGlasner commented on static/js/TopicPage.jsx:
```diff
@@ -60,16 +60,19 @@ const fetchBulkText = (translationLanguagePreference, inRefs) =>
-      if (outSheets[tempSheet.ref]) {
-        outSheets[tempSheet.ref].order = tempSheet.order;
+const fetchBulkSheet = inSheets => {
+   inSheets.forEach(x => {
```
add space for right indent

YishaiGlasner commented on static/js/TopicPage.jsx:
```diff
@@ -60,16 +60,19 @@ const fetchBulkText = (translationLanguagePreference, inRefs) =>
+   inSheets.forEach(x => {
+      x.ref = parseInt(x.ref.replace('Sheet ', ''));
+    });
+    return Sefaria.getBulkSheets(inSheets.map(x => x.ref)).then(outSheets => {
```
it's the second loop on inSheets. you can do only one?

YishaiGlasner commented on static/js/TopicPage.jsx:
```diff
@@ -563,8 +567,100 @@ const getTopicPageAnalyticsData = (slug, langPref) => {
     };
 };
 
+const setupDisplayTabs = (topic, currentTab, tabDisplayData, refsToFetchByTab, topicData, setLoadedData) => {
```
the diff here is to difficult for me to follow.
the existing code is terrible, and this tries to improve it, but on the one hand, this code is still not good enough for simply understanding it, and on the other hand it's too different from the original for understanding what was done.
@yitzhakc do you have an dea how to review in such a case?

YishaiGlasner commented on static/js/sefaria/sefaria.js:
```diff
@@ -2871,27 +2871,37 @@ _media: {},
         let tabKey = linkTypeSlug;
-        if (tabKey === 'about') {
+        let { title } = linkTypeObj;
+        if (tabKey === 'about') {  // "Sheets", "Sources", and "Notable Sources" are tagged 'about' in the DB so we have to differentiate between them here
```
all of these conitions are difficult to read.
here is gpt suggestion. it's more readable, but i didn't check it it really does the same thing... :

if (tabKey === 'about') {
  const desc = refObj.descriptions?.[lang];
  const isNotableSource = (desc?.title || desc?.prompt) && desc?.published !== false;

  if (Sefaria.activeModule === 'sheets') {
    tabKey = 'sheets';
    title = { en: 'Sheets', he: Sefaria.translation('hebrew', 'Sheets') };
  } 
  else if (isNotableSource) {
    tabKey = 'notable-sources';
    title = { en: 'Notable Sources', he: Sefaria.translation('hebrew', 'Notable Sources') };
  } 
  else {
    tabKey = 'sources';
    title = { en: 'Sources', he: Sefaria.translation('hebrew', 'Sources') };
  }
}

stevekaplan123 commented:
@yitzhakc @YishaiGlasner 
In sefaria.js, I renamed processTopicsData to processTopicsTabsData since that's more descriptive of its purpose.  I added comments and added helper functions inside processTopicsTabsData so that hopefully it's more clear and organized.  
In TopicPage.jsx, I divided up the helper function setupDisplayTabs into two functions -- one for adding tabs that display sources and one for miscellaneous tabs and added comments.  Hope it's more clear this time!

YishaiGlasner commented on static/css/s2.css:
```diff
@@ -2783,8 +2783,8 @@ a.cardTitle:hover {
 }
-.interface-english .navBlockDescription.clamped .int-en,
-.interface-hebrew .navBlockDescription.clamped .int-he {
+.interface-english .navBlockDescription.clamped .int-en, .interface-hebrew .navBlockDescription.clamped .int-he, .cardDescription
```
i meant that our convention is no new line for the opening curl:
.interface-english .navBlockDescription.clamped .int-en,
.interface-hebrew .navBlockDescription.clamped .int-he,
.cardDescription {

YishaiGlasner replied:
it still there

YishaiGlasner commented:
> @yitzhakc @YishaiGlasner In sefaria.js, I renamed processTopicsData to processTopicsTabsData since that's more descriptive of its purpose. I added comments and added helper functions inside processTopicsTabsData so that hopefully it's more clear and organized. In TopicPage.jsx, I divided up the helper function setupDisplayTabs into two functions -- one for adding tabs that display sources and one for miscellaneous tabs and added comments. Hope it's more clear this time!

ok. it seems i can't do it in a reasonable time. it looks better than it was, but i didn't know this code and this pr is not the best way to know it for the first time.

stevekaplan123 commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,90 @@ _media: {},
+            : { tabKey: 'sources', title: { en: 'Sources', he: Sefaria.translation('hebrew', 'Sources') } };
         }
+        if (linkType === 'popular-writing-of') {
+          return { tabKey: linkType, title: topicLinks.title };  // in this case the title is already set to "Top Citations"
```
This is great.  Thank you CoPilot!

akiva10b commented on static/js/TopicPage.jsx:
```diff
@@ -61,16 +61,20 @@ const fetchBulkText = (translationLanguagePreference, inRefs) =>
-const fetchBulkSheet = inSheets =>
-    Sefaria.getBulkSheets(inSheets.map(x => x.ref)).then(outSheets => {
+const fetchBulkSheet = inSheets => {
+  const refs = inSheets.map(x => {
```
add a comment that getBulkSheets expects full refs without "sheets "

akiva10b reviewed (commented):
have a look at other tab views and see if there are any reusable components here

yitzhakc commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
+        if (linkType === 'about') {
+          const lang = Sefaria.interfaceLang === "hebrew" ? 'he' : 'en';
+          const desc = refObj.descriptions?.[lang];
+          const isNotableSource = (desc?.title || desc?.prompt) && desc?.published !== false;
```
I don't love this - the frontend code shouldn't need such complicated logic to figure out if it's a notable source or not - it should be part of the data...
@yonadavGit any thoughts?

stevekaplan123 replied:
@yitzhakc I totally agree with you.  The bizarre linkType "about" should be divided into two types, but it is currently referenced in many places in our backend code, so the necessary changes would be a fairly significant re-factor. I think it should be its own story.

yitzhakc commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
-        if (tabKey === 'about') {
-            // Mark as notable if it has a prompt, title, and isn't explicitly unpublished (happens when returns from LLM pod);
-            tabKey = (refObj.descriptions?.[lang]?.title || refObj.descriptions?.[lang]?.prompt) && refObj.descriptions?.[lang]?.published != false ? 'notable-sources' : 'sources';
+  _createTabKeyAndTitle: function (linkType, refObj, lang) {
```
I'd ask that you change the name to `_getTabKey` - since ultimately, we are not creating the tab here (that happens in `processTopicsTabsData`), we're just "divining" it out based on the data provided.

yitzhakc commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
+          if (!tabKey) continue;
+          if (!tabs[tabKey]) {
+          tabs[tabKey] = {
+              refs: new Set(),
```
Very confusing that you have `refs` here as a key in this object, and also a `refs` variable (which should probably be a `const`) on line 2901. Please change the name of one of them.

yitzhakc commented:
> > have a look at other tab views and see if there are any reusable components here
> 
> I don't see how to immediately make TopicPageTabView work based on purely reusable components, although I noticed a few things that could be cleaned up. First of all, every client of TabView creates an array 'tabs' which is just a list of javascript objects and the client also passes a renderTab function to TabView. These renderTab functions are very similar to each other and it seems like we should just have a component Tab that handles the data passed to the array and also can render itself. I also found one glaring issue that entangles TabView (which we use many times on the site) to TopicPageTabView. Strangely, the filter button and lang toggle for TopicPageTabView are considered tabs which makes no sense conceptually and also entangles the two components together more than they need to be. In the CollectionsPage we also consider filter to be a tab so that component could also be cleaned up by separating filter/lang toggles out of TabView.
> 
> @yitzhakc

I agree @stevekaplan123, I would love for us to turn the filter button and language toggle into separate components and not tabs. But I don't think it should be done in this PR, we should create a separate story and PR so as not to break the scope here.

YishaiGlasner commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
+      if (refObj.is_sheet) {
+        return { tabKey: 'sheets', title: { en: "Sheets", he: Sefaria.translation('hebrew', "Sheets") } };
+      }
+      else {
```
i think the convention in js is `else if (...) {...}`. you don't need {} for the else itself.

yitzhakc replied:
@YishaiGlasner it looks like there are two `if`s inside the `else` block, so @stevekaplan123 would still need a block - but I agree that the `else` doesn't need to be on a new line :-)

YishaiGlasner commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
+      }
+      else {
+        if (linkType === 'about') {
+          const lang = Sefaria.interfaceLang === "hebrew" ? 'he' : 'en';
```
i think `Sefaria.interfaceLang.slice(0, 2)` is more readable.

yitzhakc replied:
@YishaiGlasner I don't think that's so readable either 🙈  

@stevekaplan123 maybe better to add a helper function to `Sefaria.js` that returns `he` or `en`? I.e that encapsulates this logic? It feels like we have it in lots of places...

stevekaplan123 replied:
@YishaiGlasner @yitzhakc I'm creating a helper function Sefaria._getShortInterfaceLang() 
How does this sound?
It can replace 15-20 cases throughout the site

YishaiGlasner commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
-        if (tabKey === 'about') {
-            // Mark as notable if it has a prompt, title, and isn't explicitly unpublished (happens when returns from LLM pod);
-            tabKey = (refObj.descriptions?.[lang]?.title || refObj.descriptions?.[lang]?.prompt) && refObj.descriptions?.[lang]?.published != false ? 'notable-sources' : 'sources';
+  _deriveTabDataForTopicLink: function (linkType, refObj) {
```
it seems that both functions get some very complicated data and doing something with it. it's not readable (i don't blame, i guess this is due the data and code state, but i can't understand what's going on here).

yitzhakc replied:
@stevekaplan123 I think add a comment explaining what this helper function is doing and what it returns.

yitzhakc commented on static/js/TopicPage.jsx:
```diff
@@ -543,24 +548,19 @@ const getTopicPageAnalyticsData = (slug, langPref) => {
     const scrollableElement = useRef();
     const clearAndSetTopic = (topic, topicTitle) => {setTopic(topic, topicTitle)};
+    const classStr = classNames({topicPanel: 1, readerNavMenu: 1});
+    const [langPref, setLangPref] = useState(Sefaria.interfaceLang);
```
Out of curiosity, why?

yitzhakc commented on static/js/sefaria/sefaria.js:
```diff
@@ -2864,72 +2863,94 @@ _media: {},
+        const activeRefs = topicLinks.refs.filter(refObj => (Sefaria.activeModule === 'sheets') === refObj.is_sheet);
+        for (let refObj of activeRefs) {
+          const {tabKey, title} = Sefaria._deriveTabDataForTopicLink(linkType, refObj);
+          if (!tabKey) continue;
```
Do we want a more robust error handling for missing tab data? Should we be reporting/logging it? What are the scenarios that we could reach such a situation?

stevekaplan123 replied:
This would only happen if there was an error in the data, though we have a line that explicitly sets this on the backend:
` refLinkType = 'popular-writing-of' if getattr(topic, 'subclass', '') == 'author' else 'about'`

yitzhakc commented on static/js/TopicPage.jsx:
```diff
@@ -464,8 +468,9 @@ const AuthorIndexItem = ({
 
-const useTabDisplayData = (translationLanguagePreference) => {
-  const getTabDisplayData = useCallback(() => [
+const useAllPossibleSourceTabs = (translationLanguagePreference) => {
```
This is a render function?

stevekaplan123 replied:
Are you referring to useAllPossibleSourceTabs?  It's a hook so that we know given a key in a particular topic.tabs, how do we fetch, sort, filter, and render the data.

stevekaplan123 commented on static/js/TopicPage.jsx:
```diff
@@ -799,6 +641,181 @@ TopicPage.propTypes = {
   topicTestVersion:    PropTypes.string
 };
 
+const TopicPageTabView = ({topic, topicData, tab, setTab, translationLanguagePreference, versionPref,
```
The logic hasn't changed.  I have only re-organized the code.
